contracts/src/libraries/xChain/BaseT1XChainReader.sol

```diff
@@ -0,0 +1,198 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.25;
+
+import { TypeCasts } from "@hyperlane-xyz/libs/TypeCasts.sol";
+import { OwnableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import { ReentrancyGuardUpgradeable } from "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+
+import { IT1Messenger } from "../IT1Messenger.sol";
+import { T1XChainMessage } from "./T1XChainMessage.sol";
+import { IT1XChainReaderCallback } from "./IT1XChainReaderCallback.sol";
+
+/**
+ * @title T1XChainReader
+ * @notice Facilitates reading data from contracts on other chains through t1
+ */
+abstract contract BaseT1XChainReader is OwnableUpgradeable, ReentrancyGuardUpgradeable {
+    // ============ Events ============
+
+    /**
+     * @notice Emitted when a cross-chain read request is made
+     * @param requestId Unique identifier for the request
+     * @param destinationDomain Domain ID of the target chain
+     * @param targetContract Address of the contract to read from
+     * @param minBlock the minimum block on the target chain that you will accept the read to be executed
+     * @param callData The encoded function call
+     * @param callback Address that will receive the response
+     */
+    event ReadRequested(
+        bytes32 indexed requestId,
+        uint32 indexed destinationDomain,
+        address targetContract,
+        uint64 minBlock,
+        bytes callData,
+        address indexed callback
+    );
+
+    /**
+     * @notice Emitted when a cross-chain read response is received
+     * @param requestId Unique identifier for the original request
+     * @param result The result data from the read operation
+     */
+    event ReadResult(bytes32 indexed requestId, bytes result);
+
+    // ============ State Variables ============
+
+    IT1Messenger public immutable messenger;
+    /// @notice The T1 messenger contract used for cross-chain communication
+
+    /// @notice The T1 prover
+    address public immutable prover;
+
+    /// @notice The local domain ID
+    uint32 public immutable localDomain;
+
+    /// @notice Maps request IDs to their callback addresses
+    mapping(bytes32 => address) public callbacks;
+
+    struct ReadRequest {
+        uint32 destinationDomain;
+        address targetContract;
+        uint64 minBlock;
+        bytes callData;
+        address callback;
+    }
+
+    // ============ Errors ============
+
+    error OnlyProver();
+    error OnlyCounterpart();
+    error InvalidCallback();
+    error ZeroAddress();
+
+    // ============ Modifiers ============
+
+    modifier onlyProver() {
```

Comment:
Please, add a little test to check this `onlyProver` modifier that is a feature part of the PR's description.

Reply:
done